File: demo/src/scala/tuiexamples/ParagraphExample.scala
package tuiexamples

import tui._
import tui.crossterm.CrosstermJni
import tui.widgets.{BlockWidget, ParagraphWidget}

import java.time.{Duration, Instant}
import scala.math.Ordering.Implicits._

object ParagraphExample {
  case class App(var scroll: Int = 0) {
    def on_tick(): Unit = {
      scroll += 1
      scroll %= 10
    }
  }

  def main(args: Array[String]): Unit = withTerminal { (jni, terminal) =>
    // create app and run it
    val tick_rate = Duration.ofMillis(250)
    val app = App()

    run_app(terminal, app, tick_rate, jni)
  }

  def run_app(
      terminal: Terminal,
      app: App,
      tick_rate: Duration,
      jni: CrosstermJni
  ): Unit = {
    var last_tick = Instant.now()

    def elapsed = java.time.Duration.between(last_tick, java.time.Instant.now())

    def timeout = {
      val timeout = tick_rate.minus(elapsed)
      new tui.crossterm.Duration(timeout.toSeconds, timeout.getNano)
    }

    while (true) {
      terminal.draw(f => ui(f, app))

      if (jni.poll(timeout)) {
        jni.read() match {
          case key: tui.crossterm.Event.Key =>
            key.keyEvent.code match {
              case char: tui.crossterm.KeyCode.Char if char.c() == 'q' => return
              case _                                                   => ()
            }
          case _ => ()
        }
      }
      if (elapsed >= tick_rate) {
        app.on_tick()
        last_tick = Instant.now()
      }
    }
  }

  def ui(f: Frame, app: App): Unit = {
    // Words made "loooong" to demonstrate line breaking.
    val s = "Veeeeeeeeeeeeeeeery    loooooooooooooooooong   striiiiiiiiiiiiiiiiiiiiiiiiiing.   "
    val long_line = s.repeat(f.size.width / s.length + 4) + "\n"

    val block = BlockWidget(style = Style(bg = Some(Color.White), fg = Some(Color.Black)))
    f.renderWidget(block, f.size)

    val text = Text.fromSpans(
      Spans.nostyle("This is a line "),
      Spans.styled("This is a line   ", Style.DEFAULT.fg(Color.Red)),
      Spans.styled("This is a line", Style.DEFAULT.bg(Color.Blue)),
      Spans.styled("This is a longer line", Style.DEFAULT.addModifier(Modifier.CROSSED_OUT)),
      Spans.styled(long_line, Style.DEFAULT.bg(Color.Green)),
      Spans.styled("This is a line", Style.DEFAULT.fg(Color.Green).addModifier(Modifier.ITALIC))
    )

    def create_block(title: String): BlockWidget =
      BlockWidget(
        borders = Borders.ALL,
        style = Style(bg = Some(Color.White), fg = Some(Color.Black)),
        title = Some(Spans.from(Span.styled(title, Style.DEFAULT.addModifier(Modifier.BOLD))))
      )

    Layout(direction = Direction.Vertical, margin = Margin(5))(
      Constraint.Percentage(25) -> ParagraphWidget(
        text = text,
        style = Style(bg = Some(Color.White), fg = Some(Color.Black)),
        block = Some(create_block("Left, no wrap")),
        alignment = Alignment.Left
      ),
      Constraint.Percentage(25) -> ParagraphWidget(
        text = text,
        style = Style(bg = Some(Color.White), fg = Some(Color.Black)),
        block = Some(create_block("Left, wrap")),
        alignment = Alignment.Left,
        wrap = Some(ParagraphWidget.Wrap(trim = true))
      ),
      Constraint.Percentage(25) -> ParagraphWidget(
        text = text,
        style = Style(bg = Some(Color.White), fg = Some(Color.Black)),
        block = Some(create_block("Center, wrap")),
        alignment = Alignment.Center,
        wrap = Some(ParagraphWidget.Wrap(trim = true)),
        scroll = (app.scroll, 0)
      ),
      Constraint.Percentage(25) -> ParagraphWidget(
        text = text,
        style = Style(bg = Some(Color.White), fg = Some(Color.Black)),
        block = Some(create_block("Right, wrap")),
        alignment = Alignment.Right,
        wrap = Some(ParagraphWidget.Wrap(trim = true))
      )
    ).render(f.size, f.buffer)
  }
}

File: demo/src/java/tuiexamples/ParagraphExample.java
package tuiexamples;

import java.time.Duration;
import java.time.Instant;
import tui.Alignment;
import tui.Borders;
import tui.Color;
import tui.Constraint;
import tui.Direction;
import tui.Frame;
import tui.Layout;
import tui.Margin;
import tui.Modifier;
import tui.Rect;
import tui.Span;
import tui.Spans;
import tui.Style;
import tui.Terminal;
import tui.Text;
import tui.WithTerminal;
import tui.crossterm.CrosstermJni;
import tui.crossterm.Event;
import tui.crossterm.KeyCode;
import tui.widgets.BlockWidget;
import tui.widgets.ParagraphWidget;

public final class ParagraphExample {
  private ParagraphExample() {}

  public static final class App {
    public int scroll;

    public App(int scroll) {
      this.scroll = scroll;
    }

    public void onTick() {
      scroll += 1;
      scroll %= 10;
    }
  }

  public static void main(String[] args) {
    WithTerminal.apply((jni, terminal) -> {
      Duration tickRate = Duration.ofMillis(250);
      App app = new App(0);
      runApp(terminal, app, tickRate, jni);
      return null;
    });
  }

  public static void runApp(Terminal terminal, App app, Duration tickRate, CrosstermJni jni) {
    Instant[] lastTick = {Instant.now()};

    while (true) {
      terminal.draw(f -> ui(f, app));

      Duration elapsed = Duration.between(lastTick[0], Instant.now());
      Duration remaining = tickRate.minus(elapsed);
      tui.crossterm.Duration timeout =
          new tui.crossterm.Duration(remaining.toSeconds(), remaining.getNano());
      if (jni.poll(timeout)) {
        Event ev = jni.read();
        if (ev instanceof Event.Key key) {
          if (key.keyEvent().code() instanceof KeyCode.Char c && c.c() == 'q') {
            return;
          }
        }
      }
      Duration elapsed2 = Duration.between(lastTick[0], Instant.now());
      if (elapsed2.compareTo(tickRate) >= 0) {
        app.onTick();
        lastTick[0] = Instant.now();
      }
    }
  }

  public static void ui(Frame f, App app) {
    String s = "Veeeeeeeeeeeeeeeery    loooooooooooooooooong   striiiiiiiiiiiiiiiiiiiiiiiiiing.   ";
    String longLine = s.repeat(f.size.width() / s.length() + 4) + "\n";

    BlockWidget block =
        BlockWidget.empty().withStyle(Style.empty().withBg(Color.White).withFg(Color.Black));
    f.renderWidget(block, f.size);

    Layout layout =
        new Layout(
            Direction.Vertical,
            Margin.of(5),
            new Constraint[] {
              new Constraint.Percentage(25),
              new Constraint.Percentage(25),
              new Constraint.Percentage(25),
              new Constraint.Percentage(25)
            }, true);
    Rect[] chunks = layout.split(f.size);

    Text text =
        Text.fromMany(
            Spans.nostyle("This is a line "),
            Spans.styled("This is a line   ", Style.DEFAULT.withFg(Color.Red)),
            Spans.styled("This is a line", Style.DEFAULT.withBg(Color.Blue)),
            Spans.styled(
                "This is a longer line", Style.DEFAULT.withAddModifier(Modifier.CROSSED_OUT)),
            Spans.styled(longLine, Style.DEFAULT.withBg(Color.Green)),
            Spans.styled(
                "This is a line",
                Style.DEFAULT.withFg(Color.Green).withAddModifier(Modifier.ITALIC)));

    ParagraphWidget paragraph0 =
        ParagraphWidget.empty(text)
            .withStyle(Style.empty().withBg(Color.White).withFg(Color.Black))
            .withBlock(createBlock("Left, no wrap"))
            .withAlignment(Alignment.Left);
    f.renderWidget(paragraph0, chunks[0]);

    ParagraphWidget paragraph1 =
        ParagraphWidget.empty(text)
            .withStyle(Style.empty().withBg(Color.White).withFg(Color.Black))
            .withBlock(createBlock("Left, wrap"))
            .withAlignment(Alignment.Left)
            .withWrap(new ParagraphWidget.Wrap(true));
    f.renderWidget(paragraph1, chunks[1]);

    ParagraphWidget paragraph2 =
        ParagraphWidget.empty(text)
            .withStyle(Style.empty().withBg(Color.White).withFg(Color.Black))
            .withBlock(createBlock("Center, wrap"))
            .withAlignment(Alignment.Center)
            .withWrap(new ParagraphWidget.Wrap(true))
            .withScroll(new ParagraphWidget.Scroll(app.scroll, 0));
    f.renderWidget(paragraph2, chunks[2]);

    ParagraphWidget paragraph3 =
        ParagraphWidget.empty(text)
            .withStyle(Style.empty().withBg(Color.White).withFg(Color.Black))
            .withBlock(createBlock("Right, wrap"))
            .withAlignment(Alignment.Right)
            .withWrap(new ParagraphWidget.Wrap(true));
    f.renderWidget(paragraph3, chunks[3]);
  }

  private static BlockWidget createBlock(String title) {
    return BlockWidget.empty()
        .withBorders(Borders.ALL)
        .withStyle(Style.empty().withBg(Color.White).withFg(Color.Black))
        .withTitle(
            Spans.from(Span.styled(title, Style.DEFAULT.withAddModifier(Modifier.BOLD))));
  }
}
